File: ALIPHATIC's EXERCISE/BINARY SEARCH TREE/BST.dart
// Validate Binary Search Tree

// Given the root of a binary tree, determine if it is a valid binary search tree (BST).

// A valid BST is defined as follows:

// The left subtree of a node contains only nodes with keys less than the node's key.
// The right subtree of a node contains only nodes with keys greater than the node's key.
// Both the left and right subtrees must also be binary search trees.
 

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *   int val;
 *   TreeNode? left;
 *   TreeNode? right;
 *   TreeNode([this.val = 0, this.left, this.right]);
 * }
 */
class Solution {
  bool isValidBST(TreeNode? root) {
    bool isValidBSTHelper(TreeNode? root, int? min, int? max) {
      if (root == null) return true;
      if (min != null && root.val <= min) return false;
      if (max != null && root.val >= max) return false;
      return isValidBSTHelper(root.left, min, root.val) && isValidBSTHelper(root.right, root.val, max);
    }
    return isValidBSTHelper(root, null, null);

  }
}

File: ALIPHATIC's EXERCISE/BINARY SEARCH TREE/BST.js
/** Validate Binary Search Tree

Given the root of a binary tree, determine if it is a valid binary search tree (BST).

A valid BST is defined as follows:

The left subtree of a node contains only nodes with keys less than the node's key.
The right subtree of a node contains only nodes with keys greater than the node's key.
Both the left and right subtrees must also be binary search trees.
*/ 


/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function(root) {
    
        const helper = (node, lower, upper) => {
            if(node === null) return true;
            
            if(node.val <= lower || node.val >= upper) return false;
            
            return helper(node.left, lower, node.val) && helper(node.right, node.val, upper);
        }
        
        return helper(root, -Infinity, Infinity);
    
};